### DISCUSSION TITLE
Has anyone had any success with creating a questions file feeding PrivateGPT?

### DISCUSSION THREAD

**Discussion by lsimcoates** [imartinez/privateGPT#922] [General]
Hi, looking at whether it is feasible to supply private GPT with a text file which contains a list of queries, which it can then work its way through and answer all the questions, then producing an output document. 
Currently I have a lot of source data and querying can be slow, so would like to just input a number of questions and let it just run till its done. Saves me having to go check on it a lot.

**Comment by lsimcoates** [comment]
I managed to sort this btw to run multiple queries one after each other without havent to type them all in.

**Reply by chinmay29hub** [reply]
Hi @lsimcoates that's great! With this we don't have to continuously monitor it. Can you share how did you achieve this? I have a list of questions but I want them to run one after automatically without typing them manually!

**Comment by lsimcoates** [comment]
Hi, yeh I did this a while ago but I'll dig out the code. 

I changed the 
"while True:
        query = input("\nEnter a query: ")
        if query == "exit":
            break
        if query.strip() == "":
            continue"
            
To pull from a text file in the directory named query.txt - here is my full code chunk, from what I remember I had this working, but my newest version no longer utilises this so I cant remember exactly

`#!/usr/bin/env python3
from dotenv import load_dotenv
from langchain.chains import RetrievalQA
from langchain.embeddings import HuggingFaceEmbeddings
from langchain.callbacks.streaming_stdout import StreamingStdOutCallbackHandler
from langchain.vectorstores import Chroma
from langchain.llms import GPT4All, LlamaCpp
import os
import argparse
import time

load_dotenv()

embeddings_model_name = os.environ.get("EMBEDDINGS_MODEL_NAME")
persist_directory = os.environ.get('PERSIST_DIRECTORY')

model_type = os.environ.get('MODEL_TYPE')
model_path = os.environ.get('MODEL_PATH')
model_n_ctx = os.environ.get('MODEL_N_CTX')
model_n_batch = int(os.environ.get('MODEL_N_BATCH',8))
target_source_chunks = int(os.environ.get('TARGET_SOURCE_CHUNKS',4))

from constants import CHROMA_SETTINGS

def main():
    # Parse the command line arguments
    args = parse_arguments()
    embeddings = HuggingFaceEmbeddings(model_name=embeddings_model_name)
    db = Chroma(persist_directory=persist_directory, embedding_function=embeddings, client_settings=CHROMA_SETTINGS)
    retriever = db.as_retriever(search_kwargs={"k": target_source_chunks})
    # activate/deactivate the streaming StdOut callback for LLMs
    callbacks = [] if args.mute_stream else [StreamingStdOutCallbackHandler()]
    # Prepare the LLM
    match model_type:
        case "LlamaCpp":
            llm = LlamaCpp(model_path=model_path, n_ctx=model_n_ctx, n_batch=model_n_batch, callbacks=callbacks, verbose=False)
        case "GPT4All":
            llm = GPT4All(model=model_path, max_tokens=model_n_ctx, backend='gptj', n_batch=model_n_batch, callbacks=callbacks, verbose=False)
        case _default:
            # raise exception if model_type is not supported
            raise Exception(f"Model type {model_type} is not supported. Please choose one of the following: LlamaCpp, GPT4All")
        
    qa = RetrievalQA.from_chain_type(llm=llm, chain_type="stuff", retriever=retriever, return_source_documents= not args.hide_source)
    # Read questions from the file
    with open("query.txt", "r") as file:
        questions = file.readlines()

    for question in questions:
        query = question.strip()
        if query == "exit":
            break
        if query == "":
            continue

        process_question(query, qa, args)

def process_question(query, qa, args):
    # Get the answer from the chain
    print(f"Processing question: {query}")  # Add this line to check if the questions are being processed
    start = time.time()
    res = qa(query)
    answer, docs = res['result'], [] if args.hide_source else res['source_documents']
    end = time.time()

    # Print the result
    print("\n\n> Question:")
    print(query)
    print(f"\n> Answer (took {round(end - start, 2)} s.):")
    print(answer)

    # Print the relevant sources used for the answer
    for document in docs:
        print("\n> " + document.metadata["source"] + ":")
        print(document.page_content)

def parse_arguments():
    parser = argparse.ArgumentParser(description='privateGPT: Ask questions to your documents without an internet connection, '
                                                 'using the power of LLMs.')
    parser.add_argument("--hide-source", "-S", action='store_true',
                        help='Use this flag to disable printing of source documents used for answers.')

    parser.add_argument("--mute-stream", "-M",
                        action='store_true',
                        help='Use this flag to disable the streaming StdOut callback for LLMs.')

    return parser.parse_args()

if __name__ == "__main__":
    main()


`
Hope this helps, pretty sure I had a little bit of trouble with this, but chatGPT was able assist me in this

**Reply by lsimcoates** [reply]
Apologies that formatting is terrible

**Reply by avalonweblon** [reply]
I got a dumb question, which .py file I need to update the above code?

I'm really new to programming :(

**Reply by lsimcoates** [reply]
Oh, I have just realised I was using a very early version of this project and have not updated any of the files to those now available. From what I can tell now it seems to run differently, and has different scripts. So I am not sure how possible this is now :( Sorry i cant be of more help.

**Reply by chinmay29hub** [reply]
Thanks @lsimcoates for the code!!
